### PR TITLE
sidecar, query, receiver: Native histogram support

### PR DESCRIPTION
**Pull request by rabenhorst**
Add native histogram support for sidecar, query, receiver. It does not include necessary changes for query UI and query frontend. The plan is to do this separate PRs. Relates to https://github.com/thanos-io/thanos/issues/5907.

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

- Added native histogram to prompb and storepb types (including helper/conversion functions) 
- Implemented native histogram appending in receiver writer
- Implemented native histogram support for related iterators

## Verification

- Native histogram e2e tests to test query (including dedup) and write through sidecar and receiver
  - Exposed raw json message in `QueryInstant` of promclient as workaround (until https://github.com/prometheus/common/pull/417 is available)
- Added test and benchmark for native histogram for receiver write

### CODE REVIEW

File: pkg/dedup/iter.go [hunk 1]
```diff
@@ -380,32 +380,32 @@ type dedupSeriesIterator struct {
...
-	lastValue := it.lastV
+	lastFloatVal, isFloatVal := it.lastFloatVal()
 	lastUseA := it.useA
 	defer func() {
-		if it.useA != lastUseA {
+		if it.useA != lastUseA && isFloatVal {
 			// We switched replicas.
 			// Ensure values are correct bases on value before At.
-			it.adjustAtValue(lastValue)
+			it.adjustAtValue(lastFloatVal)
```
Comment (fpetkovski) [inline pkg/dedup/iter.go:408]: I am not sure if we need to adjust histograms here, it would be quite challenging to do it the same way we adjust counters due to dynamic buckets. Let's maybe add a `TODO` and we can investigate later on.

File: docs/proposals-accepted/202301-distributed-query-execution.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+* Thanos users who have challenges with evaluating PromQL queries due to high cardinality.
+
+## 5 Goals
+
+* Enable decentralized query execution by delegating query plan fragments to independent Queriers.
+
+## 6 Proposal
+
+The key advantage of distributed execution is the fact that the number of series is drastically reduced when a query contains an aggregation operator (`sum`, `group`, `max`, etc..). Most (if not all) high cardinality PromQL queries are in-fact aggregations since users will struggle to sensibly visualise more than a handful of series.
```
Comment (yeya24) [inline docs/proposals-accepted/202301-distributed-query-execution.md:40]: Can we rebase main so that unrelated changes will not be included in this pr?

File: cmd/thanos/receive.go [hunk 3]
```diff
@@ -88,6 +88,8 @@ func registerReceive(app *extkingpin.App) {
 			EnableExemplarStorage:          true,
 			HeadChunksWriteQueueSize:       int(conf.tsdbWriteQueueSize),
 			EnableMemorySnapshotOnShutdown: conf.tsdbMemorySnapshotOnShutdown,
+			//TODO (rabenhorst): Do we want to make this configurable?
+			EnableNativeHistograms: true,
```
Comment (rabenhorst) [inline cmd/thanos/receive.go:92]: Should we put this behind a flag?

Comment (fpetkovski) [reply cmd/thanos/receive.go:92]: We can put it behind a hidden flag until we have full support.

File: pkg/store/prometheus.go [hunk 4]
```diff
@@ -255,7 +255,7 @@ func (p *PrometheusStore) queryPrometheus(s storepb.Store_SeriesServer, r *store
 		}
 		matrix = result
 	} else {
-		vector, _, err := p.client.QueryInstant(s.Context(), p.base, r.ToPromQL(), timestamp.Time(r.MaxTime), opts)
+		vector, _, _, err := p.client.QueryInstant(s.Context(), p.base, r.ToPromQL(), timestamp.Time(r.MaxTime), opts)
```
Comment (fpetkovski) [inline pkg/store/prometheus.go:258]: Hm, are we dropping a potential histogram response here? Is it possible to return a histogram?

**Review by fpetkovski — COMMENTED** [review]
Looks good, thanks. Just one comment from my side

File: pkg/store/prometheus.go (hunk 4 above)
Comment (rabenhorst) [reply pkg/store/prometheus.go:258]: This is just the raw json message exposed as a workaround until https://github.com/prometheus/common/pull/417 is available, so we can check if response contains histogram: https://github.com/rabenhorst/thanos/blob/8edbf720be226a7df48eb49a10c03d792452dd53/pkg/promclient/promclient.go#L388

File: test/e2e/native_histograms_test.go [hunk 5]
```diff
@@ -28,13 +32,15 @@ func TestQueryNativeHistograms(t *testing.T) {
 	testutil.Ok(t, err)
 	t.Cleanup(e2ethanos.CleanScenario(t, e))
 
-	prom, sidecar := e2ethanos.NewPrometheusWithSidecar(e, "prom", e2ethanos.DefaultPromConfig("prom-alone", 0, "", "", e2ethanos.LocalPrometheusTarget), "", "quay.io/prometheus/prometheus:v2.40.5", "", "native-histograms", "remote-write-receiver")
-	testutil.Ok(t, e2e.StartAndWaitReady(prom, sidecar))
+	prom1, sidecar1 := e2ethanos.NewPrometheusWithSidecar(e, "ha1", e2ethanos.DefaultPromConfig("prom-ha", 0, "", "", e2ethanos.LocalPrometheusTarget), "", "quay.io/prometheus/prometheus:v2.40.5", "", "native-histograms", "remote-write-receiver")
```
Comment (yeya24) [inline test/e2e/native_histograms_test.go:35]: Can we set our default Prometheus image to `quay.io/prometheus/prometheus:v2.40.5` or newer?

File: pkg/query/querier_test.go [hunk 6]
```diff
@@ -843,7 +843,6 @@ func (s *mockedSeriesIterator) At() (t int64, v float64) {
 	return sample.t, sample.v
 }
 
-// TODO(rabenhorst): Needs to be implemented for native histogram support.
 func (s *mockedSeriesIterator) AtHistogram() (int64, *histogram.Histogram) {
```
Comment (yeya24) [inline pkg/query/querier_test.go:846]: I see we still panic in this method but removed the comment. Do we still need to implement anything?

File: cmd/thanos/receive.go [hunk 7]
```diff
@@ -895,6 +897,10 @@ func (rc *receiveConfig) registerFlag(cmd extkingpin.FlagClause) {
 		"[EXPERIMENTAL] Enables feature to snapshot in-memory chunks on shutdown for faster restarts.").
 		Default("false").Hidden().BoolVar(&rc.tsdbMemorySnapshotOnShutdown)
 
+	cmd.Flag("tsdb.enable-native-histograms",
```
Comment (yeya24) [inline cmd/thanos/receive.go:900]: For query and sidecar, do you need to have this flag or native histograms are supported by automatically?

Comment (rabenhorst) [reply cmd/thanos/receive.go:900]: This flag is only to enable native histograms for the receiver tsdb: https://github.com/thanos-io/thanos/blob/b8d6b12e24314cf343b56290625ca1c60ef1eae8/cmd/thanos/receive.go#L91

For query and sidecar it's supported by automatically.

File: test/e2e/native_histograms_test.go (hunk 5 above)
Comment (rabenhorst) [reply test/e2e/native_histograms_test.go]: I updated it to v2.41.0.

**Comment by rabenhorst** [comment]
I updated prometheus/common so that it includes https://github.com/prometheus/common/pull/417, which is not tagged yet but helps to improve tests (see last commit).

File: pkg/query/querier_test.go (hunk 6 above)
Comment (rabenhorst) [reply pkg/query/querier_test.go:846]: Will restore it. Rn `mockedSeriesIterator` does not support histograms and if we need it it needs to be implemented.

**Review by fpetkovski — APPROVED** [review]
Thanks, this looks good now 👍

**Review by fpetkovski — APPROVED** [review]

File: pkg/dedup/iter.go [hunk 8]
```diff
@@ -380,32 +380,33 @@ type dedupSeriesIterator struct {
...
 	}
 }
 
 func (it *dedupSeriesIterator) Next() chunkenc.ValueType {
-	lastValue := it.lastV
+	lastFloatVal, isFloatVal := it.lastFloatVal()
 	lastUseA := it.useA
 	defer func() {
-		if it.useA != lastUseA {
+		if it.useA != lastUseA && isFloatVal {
```
Comment (GiedriusS) [inline pkg/dedup/iter.go:405]: `adjustAtValue` seems to already be protected against doing something with non-float values but this makes code clearer :+1:

**Review by GiedriusS — COMMENTED** [review]
LGTM, just a nit

File: pkg/query/iter.go [hunk 9]
```diff
@@ -289,24 +292,24 @@ func (it *chunkSeriesIterator) At() (t int64, v float64) {
...
+	return it.chunks[it.i].AtT()
 }
 
 func (it *chunkSeriesIterator) Next() chunkenc.ValueType {
-	lastT, _ := it.At()
+	lastT := it.AtT()
 
-	if valueType := it.chunks[it.i].Next(); valueType != chunkenc.ValNone {
+	var valueType chunkenc.ValueType
+	if valueType = it.chunks[it.i].Next(); valueType != chunkenc.ValNone {
```
[the remark this answers is not in the document]
Comment (rabenhorst) [reply pkg/query/iter.go]: Done!

**Review by GiedriusS — APPROVED** [review]
